### ISSUE TITLE
xrdp connects to VNC on 127.0.0.1 (v4) with IPV6_V6ONLY configured on client socket, which fails

### ISSUE THREAD

**Issue by TLINDEN** [neutrinolabs/xrdp#1248]
I am using FreeBSD 11.2, with xrdp xrdp-0.9.8,1 from Ports.  I've got an Xvnc running on localhost:5968:

`Xvnc :68 -desktop X -httpd /usr/local/share/tightvnc/classes -auth /home/tom/.Xauthority -geometry 1600x1100 -depth 24 -rfbwait 120000 -rfbauth /home/tom/.vnc/passwd -rfbport 5968 -compatiblekbd -localhost`

It really listens on localhost w/ tcp4:
`tcp4       0      0 127.0.0.1.5968         *.*                    LISTEN`

It is possible to connect by using ipv4 telnet or `vncviewer`:
```
# telnet 127.0.0.1 5986
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
RFB 003.008
^D
```

Xrdp config is pretty standard, here's the relevant vnc section for my user:
```
[tom]   
name=Tom
lib=libvnc.so
ip=127.0.0.1
port=5968
username=na
password=ask
```

When logging in via xrdp, I see the following in the logs:
```
[20181114-08:36:18] [DEBUG] xrdp_wm_log_msg: VNC started connecting
[20181114-08:36:18] [DEBUG] xrdp_wm_log_msg: VNC connecting to 127.0.0.1 5968
[20181114-08:36:21] [DEBUG] VNC error 1 after security negotiation
[20181114-08:36:21] [DEBUG] VNC error before sending share flag
[20181114-08:36:21] [DEBUG] VNC error before receiving server init
[20181114-08:36:21] [DEBUG] VNC error before receiving pixel format
[20181114-08:36:21] [DEBUG] VNC error before receiving name length
[20181114-08:36:21] [DEBUG] VNC error before receiving name
[20181114-08:36:21] [DEBUG] xrdp_wm_log_msg: VNC error - problem connecting
[20181114-08:36:21] [DEBUG] Closed socket 17 (AF_INET6 ::1 port 42298)
[20181114-08:36:21] [DEBUG] xrdp_wm_log_msg: some problem
[20181114-08:36:21] [ERROR] xrdp_wm_log_msg: Error connecting to: 127.0.0.1
[20181114-08:36:21] [DEBUG] xrdp_mm_module_cleanup
[20181114-08:36:21] [DEBUG] VNC mod_exit
[20181114-08:36:22] [DEBUG] return value from xrdp_mm_connect 1
```

I traced the xrdp process and this is what it does (snipped to the relevant part):
```
socket(PF_INET6,SOCK_STREAM,0)                   = 17 (0x11)
getsockopt(17,IPPROTO_IPV6,IPV6_V6ONLY,0x7fffdfdfb474,0x7fffdfdfb470) = 0 (0x0)
setsockopt(17,IPPROTO_IPV6,IPV6_V6ONLY,0x7fffdfdfb474,4) = 0 (0x0)
getsockopt(17,SOL_SOCKET,SO_REUSEADDR,0x7fffdfdfb474,0x7fffdfdfb470) = 0 (0x0)
setsockopt(17,SOL_SOCKET,SO_REUSEADDR,0x7fffdfdfb474,4) = 0 (0x0)
getsockopt(17,SOL_SOCKET,SO_SNDBUF,0x7fffdfdfb474,0x7fffdfdfb470) = 0 (0x0)
fcntl(17,F_GETFL,)                               = 2 (0x2)
fcntl(17,F_SETFL,O_RDWR|O_NONBLOCK)              = 0 (0x0)
connect(17,{ AF_INET6 [::1]:5968 },28)           ERR#36 'Operation now in progress'
nanosleep({ 0.600000000 })                       = 0 (0x0)
connect(17,{ AF_INET6 [::1]:5968 },28)           ERR#61 'Connection refused'
connect(17,{ AF_INET 127.0.0.1:5968 },16)        ERR#22 'Invalid argument'
connect(17,{ AF_INET6 [::ffff:127.0.0.1]:5968 },28) ERR#61 'Connection refused'
nanosleep({ 0.600000000 })                       = 0 (0x0)
connect(17,{ AF_INET6 [::1]:5968 },28)           ERR#61 'Connection refused'
connect(17,{ AF_INET 127.0.0.1:5968 },16)        ERR#22 'Invalid argument'
connect(17,{ AF_INET6 [::ffff:127.0.0.1]:5968 },28) ERR#61 'Connection refused'
nanosleep({ 0.600000000 })                       = 0 (0x0)
connect(17,{ AF_INET6 [::1]:5968 },28)           ERR#61 'Connection refused'
connect(17,{ AF_INET 127.0.0.1:5968 },16)        ERR#22 'Invalid argument'
connect(17,{ AF_INET6 [::ffff:127.0.0.1]:5968 },28) ERR#61 'Connection refused'
nanosleep({ 0.600000000 })                       = 0 (0x0)
connect(17,{ AF_INET6 [::1]:5968 },28)           ERR#61 'Connection refused'
connect(17,{ AF_INET 127.0.0.1:5968 },16)        ERR#22 'Invalid argument'
connect(17,{ AF_INET6 [::ffff:127.0.0.1]:5968 },28) ERR#61 'Connection refused'
nanosleep({ 0.600000000 })                       = 0 (0x0)
connect(17,{ AF_INET6 [::1]:5968 },28)           ERR#61 'Connection refused'
connect(17,{ AF_INET 127.0.0.1:5968 },16)        ERR#22 'Invalid argument'
connect(17,{ AF_INET6 [::ffff:127.0.0.1]:5968 },28) ERR#61 'Connection refused'
```

As you can see, xrdp sets `IPV6_V6ONLY` and then tries to use v4 and v6 on that socket, which cannot work of course.

Please use separate client sockets or add an option in `xrdp.ini` where I can force it to use ipv4 (which I already did though, since I have `ip=127.0.0.1`)

Thanks,
Tom

**Comment by TLINDEN** [comment]
Update: If I setup a v6=>v4 port forwarding using `socat`:
```
socat TCP6-LISTEN:5968,reuseaddr,fork,su=nobody TCP4:127.0.0.1:5968
```
I can connect.

So, I'd consider this a bug.

**Comment by metalefty** [comment]
Thanks for reporting, it should be a bug. I'll look at it.

**Comment by TLINDEN** [comment]
I updated to the latest available port: xrdp-0.9.10_1,1 but the problem persists.

**Comment by TLINDEN** [comment]
Oh, and there's another issue. As I already mentioned earlier I am able to connect if I run a socat v6=>v4 proxy. When I disconnect I can see this in the log:
```
[20190508-09:38:58] [DEBUG] Closed socket 12 (AF_INET6 ::ffff:172.20.131.21 port 3389)
```

`::ffff:172.20.131.21` seems to be the source address of the client (that v4 part is indeed the ip of the system from where I came via socat), which is no prober v4 nor a proper v6 address.

**Comment by matt335672** [comment]
This is all fixed in v0.10.x which removes the need to communicate with sesman over a TCP socket.
